process_pr.py

```diff
@@ -1466,86 +1470,194 @@ def process_pr(repo_config, gh, repo, issue, dryRun, cmsbuild_user=None, force=F
+                    events[datetime.fromtimestamp(cache_entry["time"])].append(
+                        {
+                            "type": "commit",
+                            "value": {"files": cache_entry["files"], "sha": commit_sha},
+                        }
+                    )
+
     # Process commits and signatures
+    signed_commit_sha = ""
+    flattened_eventlist = []
     for _, eventlist in sorted(events.items()):
-        for event in eventlist:
-            print("Event:", event)
-            if event["type"] == "sign":
-                selected_cats = event["value"]["selected_cats"]
-                ctype = event["value"]["ctype"]
-                if any(x in signing_categories for x in selected_cats):
-                    set_comment_emoji_cache(
-                        dryRun, bot_cache, event["value"]["comment"], repository
+        flattened_eventlist.extend(eventlist)
+
+    for event in flattened_eventlist:
+        print("Event:", event)
+        if event["type"] == "sign":
+            comment = event["value"]["comment"]
+            cached_signed_commit_sha = bot_cache["signatures"].get(comment.id)
+            if cached_signed_commit_sha and cached_signed_commit_sha != signed_commit_sha:
+                print(
+                    "WARNING: For comment {0}, cached commit's SHA doesn't match the present list of signed commits. This comment will be ignored."
+                )
+                continue
+
+            if cached_signed_commit_sha is None:
+                bot_cache["signatures"][comment.id] = signed_commit_sha
+            selected_cats = event["value"]["selected_cats"]
+            ctype = event["value"]["ctype"]
+            if any(x in signing_categories for x in selected_cats):
+                set_comment_emoji_cache(dryRun, bot_cache, event["value"]["comment"], repository)
+
+                if ctype == "+1":
+                    for sign in selected_cats:
+                        signatures[sign] = "approved"
+                        if (
+                            issue.pull_request
+                            and (test_comment is None)
+                            and ((repository in auto_test_repo) or ("*" in auto_test_repo))
+                            and sign not in ("code-checks", "tests", "orp")
+                            and (comment.created_at >= last_commit_date)
+                        ):
+                            test_comment = comment
+                elif ctype == "-1":
+                    for sign in selected_cats:
+                        signatures[sign] = "rejected"
+            else:
+                print(
+                    "Ignoring event: {0} includes none of {1}".format(
+                        selected_cats, signing_categories
+                    )
+                )
+        elif event["type"] == "commit":
+            if event["value"]["files"]:
+                signed_commit_sha = event["value"]["sha"]
```

Comment:
@iarspider , I think the condition `if event["value"]["files"]` is not needed. ` signed_commit_sha = event["value"]["sha"]` should work for all commits events

Reply:
Do we want to sign squash/merge commits? This check is there to skip such commits.